### PR TITLE
Add taxonomy function

### PR DESCRIPTION
**Pull request by damianooldoni**
This PR :
1. adds `taxonomy()` function (#40) with related documentation
2. adds tests for `taxonomy()`
3. joins output of `build_taxonomy()` to observations by `scientificName` (#17). These extra columns will be identified by prefix `"taxon."`
4. adds tests for `observations()` about the added taxonomic information
5. adds `.data` to `camtrapdp-package.R` to use it everywhere in the package as done for `%>%` (#37)
6. updates `NEWS.md`

Notice that there are still tests where `dataset$data$observations <- ` is used. To be replaced by `observations(dataset) <-` when this feature is implemented (#42) .

### CODE REVIEW

**Comment by PietrH** [comment]
#28 was merged

**Comment by PietrH** [comment]
I'm getting a R CMD Check note: 

```
N  checking R code for possible problems (4.1s)
   read_camtrapdp: no visible binding for global variable ‘scientificName’
   read_camtrapdp: no visible binding for global variable
     ‘taxon.scientificName’
   Undefined global functions or variables:
     scientificName taxon.scientificName
```

This line is the culprit: 

https://github.com/inbo/camtrapdp/blob/fc5126460b07f858239808042fbb4497eccb3b83/R/read_camtrapdp.R#L53

**Comment by damianooldoni** [comment]
> I'm getting a R CMD Check note:
> 
> ```
> N  checking R code for possible problems (4.1s)
>    read_camtrapdp: no visible binding for global variable ‘scientificName’
>    read_camtrapdp: no visible binding for global variable
>      ‘taxon.scientificName’
>    Undefined global functions or variables:
>      scientificName taxon.scientificName
> ```
> 
> This line is the culprit:
> 
> https://github.com/inbo/camtrapdp/blob/fc5126460b07f858239808042fbb4497eccb3b83/R/read_camtrapdp.R#L53

Solve via [bb20dd4](https://github.com/inbo/camtrapdp/pull/38/commits/bb20dd43fbb5ca482030e523d1e150f4cbc7bcac)

File: tests/testthat/test-taxonomy.R
[the remark this answers is not in the document]
Comment (PietrH) [reply tests/testthat/test-taxonomy.R:1]: We should check that taxonomy() doesn't result in observations being repeated after the `left_join`

**Review by PietrH — CHANGES_REQUESTED** [review]

File: tests/testthat/test-taxonomy.R [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  dataset <- example_dataset()
+  taxa <- taxonomy(dataset)
+  expect_true("scientificName" %in% names(taxa))
+  expect_true(
+    all(
+      grepl(pattern = "^taxon\\.",
+              x = names(taxa %>% dplyr::select(-scientificName))
+            )
+    )
+  )
```
[the remark this answers is not in the document]
Comment (damianooldoni) [reply tests/testthat/test-taxonomy.R:23]: This is actually what I had in my mind, actually. I agree with you that hard coding is the best way for writing codes.
However, @peterdesmet didn't want to hard code the column names, unless I misunderstood him?

File: tests/testthat/test-taxonomy.R [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+      distinct(scientificName) %>%
+      filter(!is.na(scientificName))
+  )
+})
+
+test_that("taxonomy returns the expected amount of rows", {
+  skip_if_offline()
+  dataset <- example_dataset()
+  taxa <- taxonomy(dataset)
+  expect_equal(nrow(taxa),10)
```
[the remark this answers is not in the document]
Comment (damianooldoni) [reply tests/testthat/test-taxonomy.R:44]: Fair enough! Precise as much as possible. Thanks.

File: tests/testthat/test-observations.R [hunk 3]
```diff
@@ -1,3 +1,6 @@
+library(purrr)
```
[the remark this answers is not in the document]
Comment (damianooldoni) [reply tests/testthat/test-observations.R:1]: I'll try to remember it.

File: tests/testthat/test-build_taxonomy.R [hunk 4]
```diff
@@ -1,4 +1,4 @@
-test_that("build_taxonomy() returns a data frame", {
+test_that("build_taxonomy() returns a tibble data frame", {
```
[the remark this answers is not in the document]
Comment (damianooldoni) [reply tests/testthat/test-build_taxonomy.R:1]: Thanks. I set it back to data frame but forgot to modify also the documentation. Good catch!

File: tests/testthat/test-taxonomy.R (hunk 1 above)
[the remark this answers is not in the document]
Comment (damianooldoni) [reply tests/testthat/test-taxonomy.R:19]: Discussed internally. Just now. We all agree on hard coding the columns 🚀

File: tests/testthat/test-taxonomy.R [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+              x = names(taxa %>% dplyr::select(-scientificName))
+            )
+    )
+  )
+})
+
+test_that(
+  "taxonomy returns scientificName column only if no taxon info in observations", {
+  skip_if_offline()
+  dataset <- example_dataset()
```
[the remark this answers is not in the document]
Comment (damianooldoni) [reply tests/testthat/test-taxonomy.R:27]: Nice idea! Any reader, even "future-me" will appreciate it for sure.

File: tests/testthat/test-taxonomy.R
[the remark this answers is not in the document]
Comment (damianooldoni) [reply tests/testthat/test-taxonomy.R:1]: Except for the columns/rows order of checks, I follow your scheme. 
Decided to first check cols and then rows.

[the remark this answers is not in the document]
Comment (damianooldoni) [reply tests/testthat/test-taxonomy.R:1]: Decided just now to not check/return warning for duplicates returned in `taxonomy()`.

File: tests/testthat/test-taxonomy.R [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+  obs <- observations(dataset)
+  taxa <- taxonomy(dataset)
+  expect_true(nrow(taxa) <= nrow(obs))
+})
+
+test_that(
+  "one row returned if all observations have same taxonomy information", {
+    skip_if_offline()
+    dataset <- example_dataset()
```
[the remark this answers is not in the document]
Comment (damianooldoni) [reply tests/testthat/test-taxonomy.R:42]: good idea.

**Comment by PietrH** [comment]
I'm going to leave this to Peter to review this time.

**Comment by damianooldoni** [comment]
Thanks @peterdesmet and @PietrH. Now, to merge this, we need an approved review.

**Review by peterdesmet — APPROVED** [review]

**Pull request merged**